feat: packfile parsing progress

johncoder opened the pull request:
This PR adds the ability for go-git to provide periodic progress updates. In particular, it includes updates for:

- Receiving objects (count, max, and percentage)
- Bytes received and rate of throughput
- Resolving deltas (count, max)

Objects and deltas are counted while scanning and parsing packfiles

It borrows heavily from the official git implementation. Some notable points:

- IEC values for human readable bytes and throughput
- Rate is calculated with a ring buffer that polls periodically. In places of timers and signals used by git, we're using goroutines.

The client provides a progress reporter, allowing them to specify the frequency of updates and how often to poll for rate calculations. The progress reporter also has a channel to read progress updates from. After the context is canceled, the consuming code ought to be sure to finish receiving all the updates.

There's also a progress collector that's used to track all the progress data points and is ultimately responsible for sending to the progress channel.

Progress is reported using a progress update struct, and for convenience implements stringer to provide human oriented output intended to closely resemble the output of git.

Commit:
wire up progress parse for clone

Commit:
update progress update string

Dynamically includes bytes received and rate

Commit:
fix: counting resolving deltas

Commit:
correctly count resolving deltas

Commit:
extract sending updates, ensure first update is sent

Commit:
citation

johncoder commented on options.go:
```diff
@@ -56,6 +57,8 @@ type CloneOptions struct {
 	// no-progress, is sent to the server to avoid send this information.
 	Progress sideband.Progress
+	// ProgressReporter is where to get local updates on received objects and resolved deltas
+	ProgressReporter *progress.ProgressReporter
```
🍜

johncoder commented on plumbing/format/packfile/common.go:
```diff
@@ -26,12 +28,28 @@ const (
+		cancel()
+	}()
+	p, err := NewParserWithStorage(NewScanner(reader, pc), s)
+	p.ProgressCollector = pc
```
🍜

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -78,15 +82,24 @@ func NewParserWithStorage(
+	}
+
+	p.ProgressCollector.ResolveDelta(p.deltasSeen, p.scanner.deltasTotal)
+}
```
🍜

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -53,6 +55,8 @@ type Parser struct {
 
 	ob []Observer
+
+	ProgressCollector *progress.ProgressCollector
```
🍜

johncoder commented on plumbing/progress/throughput.go:
```diff
@@ -0,0 +1,57 @@
+	}
+}
+
+// AdvanceTime tracks the progress and returns the current rate of bytes being read
+// SEE: https://github.com/git/git/blob/be8661a3286c67a5d4088f4226cbd7f8b76544b0/progress.c#L186-L244
+func (t *Throughput) AdvanceTime(now int64) uint64 {
+	// a "misec" is 1/1024 of a second, useful for simplifying the
+	// calculation of rate in IEC units
+	misecs := ((uint64(now) - t.PreviousNanoseconds) * 4398) >> 32
+	count := uint64(t.CurrentTotal - t.PreviousTotal)
+
+	t.PreviousTotal = t.CurrentTotal
+	t.PreviousNanoseconds = uint64(now)
+	t.AverageBytes += count
+	t.AverageMisecs += misecs
+
+	rate := t.AverageBytes / t.AverageMisecs
+
+	t.AverageBytes -= t.LastBytes[t.Index]
+	t.AverageMisecs -= t.LastMisecs[t.Index]
+	t.LastBytes[t.Index] = count
+	t.LastMisecs[t.Index] = misecs
+	t.Index = (t.Index + 1) % ThroughputRateSlotCount
+
+	return rate
+}
```
🍜

johncoder commented on plumbing/progress/progress_collector.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// Start begins a background process that will send periodic progress updates
+func (pc *ProgressCollector) Start(ctx context.Context) {
```
🍜

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -271,6 +290,12 @@ func (p *Parser) resolveDeltas() error {
+		if obj.DiskType.IsDelta() {
+			p.deltasSeen++
+			p.writeProgress()
+		}
```
🍜

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -256,8 +269,14 @@ func (p *Parser) indexObjects() error {
 		p.oiByOffset[oh.Offset] = ota
 		p.oi[i] = ota
+		p.scanner.objectsSeen++
+		p.scanner.writeProgress()
```
🍜

johncoder commented on plumbing/format/packfile/scanner.go:
```diff
@@ -50,19 +51,34 @@ type Scanner struct {
+		return
 	}
+
+	s.ProgressCollector.ReceiveObject(s.objectsSeen, s.objects)
```
🍜

davetakahashi-abstract commented on plumbing/format/packfile/common.go:
```diff
@@ -26,12 +28,28 @@ const (
+		pc.Start(ctx)
+		defer func() {
+			cancel()
+		}()
```
I'm probably missing something, but can these just be `defer cancel()`?

johncoder replied:
💯 Just some leftovers.

davetakahashi-abstract commented on plumbing/progress/progress_collector.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// ProgressCollector collects data to send progress updates
+type ProgressCollector struct {
```
Nit: my linter is complaining about package + func name "stuttering" e.g. `progress.ProgressCollector`

johncoder replied:
Good call, applied this to everything in this new progress package 👍

davetakahashi-abstract commented on plumbing/progress/progress_collector.go:
```diff
@@ -0,0 +1,154 @@
+	pr            *ProgressReporter
+}
+
+func NewProgressCollector(reader io.Reader, pr *ProgressReporter) *ProgressCollector {
```
Nit: linter is also complaining about commenting exported funcs and vars.

davetakahashi-abstract commented on plumbing/progress/progress_collector.go:
```diff
@@ -0,0 +1,154 @@
+type ProgressCollector struct {
+	receiveObject chan *CountingProgressUpdate
+	resolveDelta  chan *CountingProgressUpdate
+	accumulate    chan int
```
Is this always a byte count / byte total? Would `byteCount` or `byteTotal` be clearer?

johncoder replied:
🤔 Could be. I am predisposed to use imperative naming on IO type things... i.e.  any int sent to this channel will be accumulated. I don't feel strongly, though.

davetakahashi-abstract replied:
no biggie

davetakahashi-abstract commented on plumbing/progress/progress_collector.go:
```diff
@@ -0,0 +1,154 @@
+
+// Start begins a background process that will send periodic progress updates
+func (pc *ProgressCollector) Start(ctx context.Context) {
+	if pc == nil || pc.pr == nil {
```
For this check, if pc is nil, it will blow up with a null ref before we hit this, won't it? For the pc.pr check, would it make sense to check that in the constructor func? (check that the ProgressReporter passed in isn't nil)

johncoder replied:
Managed to get rid of these checks in favor of more sensible nil checking in the consuming code. There was some code that went something like this:
```go
var pc *progress.Collector = nil
if someCondition {
  pc = progress.NewCollector(...)
}
// some other things
pc.Start(ctx) // oops, pc is still nil
```

So the first check for `pc == nil` came from an actual nil pointer, and pretty easily avoided.

johncoder replied:
Also, I added a `*Reporter.CreateCollector` function to help guarantee the Collector has a Reporter. I could look at unexporting `NewCollector`.

davetakahashi-abstract reviewed (commented):
Leaving a couple nitty comments, should be able to dig in more tomorrow =)

Commit:
add *Reporter.CreateCollector

This isn't my favorite, but it helps guarantee that the collector will
have a progress reporter.

Added some more comments for exported things

Commit:
fix test

Commit:
remove package stutter, fix comment for exported function

johncoder commented on plumbing/format/packfile/common.go:
```diff
@@ -26,12 +28,30 @@ const (
 // UpdateObjectStorage updates the storer with the objects in the given
 // packfile.
-func UpdateObjectStorage(s storer.Storer, packfile io.Reader) error {
+func UpdateObjectStorage(s storer.Storer, packfile io.Reader, pr *progress.Reporter) error {
```
🍜

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -256,8 +279,14 @@ func (p *Parser) indexObjects() error {
 			copy(p.deltas[oh.Offset], data)
 		}
 
+		if delta {
+			p.deltasTotal++
+		}
```
🍜 We count the total of deltas in the packfile here so that we can track progress while resolving them later.

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -256,8 +279,14 @@ func (p *Parser) indexObjects() error {
+
 		p.oiByOffset[oh.Offset] = ota
 		p.oi[i] = ota
+		p.objectsSeen++
+		p.writeObjectProgress()
```
This has proven to be an effective place to count objects we're receiving, but I'm realizing that this may be more of a quick enumeration than actually representative of receiving the objects. Might give this a bit more thought.

johncoder commented on plumbing/format/packfile/parser.go:
```diff
@@ -271,6 +300,12 @@ func (p *Parser) resolveDeltas() error {
 			return err
 		}
+
+		if obj.DiskType.IsDelta() {
+			p.deltasSeen++
+			p.writeDeltaProgress()
+		}
```
🍜 The key here was that `obj.DiskType` will indicate that it's a delta, rather than using `obj.Type` (for reasons I don't really understand).

johncoder commented on plumbing/progress/collector.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+// Start begins a background process that will send periodic progress updates
+func (pc *Collector) Start(ctx context.Context) {
```
🍜

johncoder commented on plumbing/progress/throughput.go:
```diff
@@ -0,0 +1,58 @@
+
+// AdvanceTime tracks the progress and returns the current rate of bytes being read
+// SEE: https://github.com/git/git/blob/be8661a3286c67a5d4088f4226cbd7f8b76544b0/progress.c#L186-L244
+func (t *Throughput) AdvanceTime(now int64) uint64 {
```
🍜

johncoder reviewed (commented):
Re-marking points of interest for this review.

apriendeau commented on plumbing/format/packfile/packfile.go:
```diff
@@ -5,11 +5,11 @@ import (
 	"github.com/goabstract/go-git/plumbing/cache"
 	"github.com/goabstract/go-git/plumbing/format/idxfile"
 	"github.com/goabstract/go-git/plumbing/storer"
+	billy "gopkg.in/src-d/go-billy.v4"
```
What is your formatter doing? 🤔

apriendeau commented on plumbing/progress/collector.go:
```diff
@@ -0,0 +1,152 @@
+		startedResolvingDeltas := false
+		doneReceivingObjects := false
+		doneResolvingDeltas := false
+		var lastObject *CountingUpdate = nil
```
you can omit the `= nil`  here since its a pointer

Commit:
remove extraneous default value

Commit:
cleanup

Commit:
feat: packfile parsing progress

First pass of add the ability to provide progress updates while
parsing a packfile.

Commit:
feat: progress collector and receiver

For collecting the progress while enumerating objects/deltas on a
packfile as it's being parsed, as well as reporting total bytes
processed and the rate of throughput.

Commit:
cleanup

+ Refactored all the progress stuff out of scanner into parser
+ Eliminated package name stuttering
+ Fixed possible nil pointer

Commit:
undoing go fmt reordering in files that are no longer touched

Commit:
fix imports
